File: go-backend-interview-prep/06_concurrency/02_channels_basics/main.py
"""Channels basics -- Python equivalent of the Go example.

Python has no built-in channels. queue.Queue is the closest match.
It has no close() or range -- use a sentinel value (None) to signal done.
"""

import queue
import threading

SENTINEL = None  # signals "channel closed"


def main() -> None:
    # --- Example 1: basic send and receive ---
    print("=== Basic send/receive ===")
    q: queue.Queue[str] = queue.Queue()

    threading.Thread(target=lambda: q.put("hello from thread")).start()

    msg = q.get()  # blocks until item available
    print(msg)

    # --- Example 2: close + range (sentinel pattern) ---
    print("\n=== Close + range (sentinel) ===")
    nums: queue.Queue[int | None] = queue.Queue()

    def producer():
        for i in range(1, 6):
            nums.put(i)
        nums.put(SENTINEL)  # signal: no more values

    threading.Thread(target=producer).start()

    while True:
        n = nums.get()
        if n is SENTINEL:
            break
        print(f"  received: {n}")

    # --- Example 3: comma-ok pattern (not native in Python) ---
    print("\n=== Get with timeout (closest to comma-ok) ===")
    q2: queue.Queue[int] = queue.Queue()
    q2.put(42)

    try:
        v = q2.get_nowait()
        print(f"  v={v}, ok=True (value available)")
    except queue.Empty:
        print("  v=0, ok=False (queue empty)")

    try:
        v = q2.get_nowait()
        print(f"  v={v}, ok=True")
    except queue.Empty:
        print("  v=0, ok=False (queue empty)")

    # --- Example 4: queue as function return ---
    print("\n=== Queue as return value ===")
    result = compute(10, 20)
    print("  10 + 20 =", result.get())


def compute(a: int, b: int) -> queue.Queue:
    q: queue.Queue[int] = queue.Queue()
    threading.Thread(target=lambda: q.put(a + b)).start()
    return q


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/06_concurrency/02_channels_basics/main.go
package main

import "fmt"

func main() {
	// --- Example 1: basic send and receive ---
	fmt.Println("=== Basic send/receive ===")
	ch := make(chan string)

	go func() {
		ch <- "hello from goroutine"
	}()

	msg := <-ch // blocks until value is sent
	fmt.Println(msg)

	// --- Example 2: close + range ---
	fmt.Println("\n=== Close + range ===")
	nums := make(chan int)

	go func() {
		for i := 1; i <= 5; i++ {
			nums <- i
		}
		close(nums) // signal: no more values
	}()

	for n := range nums { // loops until channel is closed
		fmt.Printf("  received: %d\n", n)
	}

	// --- Example 3: comma-ok pattern ---
	fmt.Println("\n=== Comma-ok receive ===")
	ch2 := make(chan int, 1)
	ch2 <- 42
	close(ch2)

	v, ok := <-ch2
	fmt.Printf("  v=%d, ok=%v (value available)\n", v, ok)

	v, ok = <-ch2
	fmt.Printf("  v=%d, ok=%v (channel closed, zero value)\n", v, ok)

	// --- Example 4: channel as function return ---
	fmt.Println("\n=== Channel as return value ===")
	result := compute(10, 20)
	fmt.Println("  10 + 20 =", <-result)
}

// compute returns a channel that will receive the sum.
func compute(a, b int) <-chan int {
	ch := make(chan int)
	go func() {
		ch <- a + b
	}()
	return ch
}
